fix: set status code before publishing diagnostics error channel

tt-a1i opened the pull request:
## Summary

Fixes #6409

Previously, the `tracing:fastify.request.handler:error` diagnostics channel would always report `reply.statusCode` as 200 because the channel was published before the error handler had a chance to set the appropriate status code.

This fix sets the status code based on the error's `statusCode` or `status` property before publishing the diagnostics channel event, so subscribers can see the correct status code.

## Changes

- `lib/handle-request.js`: Set status code before publishing error channel (both sync error paths)
- `lib/wrap-thenable.js`: Set status code before publishing error channel (async error path)
- Added tests verifying the fix

## Test plan

- [x] Added test: `diagnostics channel error event should report correct status code` - verifies error with statusCode
- [x] Added test: `diagnostics channel error event should report 500 for errors without status` - verifies plain errors default to 500
- [x] Added test: `diagnostics channel error event should report correct status with custom error handler` - documents behavior with custom error handlers
- [x] All existing tests pass

Eomm commented on test/diagnostics-channel/error-status.test.js:
```diff
@@ -5,6 +5,99 @@ const Fastify = require('../..')
+  const res = await fastify.inject({
+    method: 'GET',
+    url: '/'
+  })
```
```suggestion
  const res = await fastify.inject('/')
```

Eomm commented on test/diagnostics-channel/error-status.test.js:
```diff
@@ -5,6 +5,99 @@ const Fastify = require('../..')
+  const res = await fastify.inject({
+    method: 'GET',
+    url: '/'
+  })
```
```suggestion
  const res = await fastify.inject('/')
```

Eomm commented on test/diagnostics-channel/error-status.test.js:
```diff
@@ -5,6 +5,99 @@ const Fastify = require('../..')
+  const res = await fastify.inject({
+    method: 'GET',
+    url: '/'
+  })
```
```suggestion
  const res = await fastify.inject('/')
```

tt-a1i commented:
Thanks for the review! I've addressed your feedback:

1. **Extracted the repeated logic** into a dedicated `setErrorStatusCode` function
2. **Simplified the inject calls** in tests using shorthand syntax

I initially tried adding the function to `Reply.js` as suggested, but it caused a circular dependency warning (`reply.js` → `handle-request.js` → `reply.js`). So I created a separate `lib/error-status.js` module instead, which follows the naming pattern of `error-handler.js` and `error-serializer.js`.

The extracted function also includes a null check (`err &&`) that was present in `error-handler.js` but missing in my original implementation - this makes the code more robust.
